### ISSUE TITLE
How to write from a blob ?

### ISSUE THREAD

**Issue by berenger-germain-sams** [qrclip/capacitor-file-chunk#2]
Hi,

I work on a ionic 6/capacitor 4/angular application that needs to download big video files (up to 500MB mp4). I already manage the download of the file through XHR which results in a blob.

I used to write it with the capacitor-blob-writer but it works only on Android with files of such size. And then I discovered your plugin. Thanks a lot for it :)

I'm not totally comfortable with blob/array buffer/etc. handling and I don't understand how to write the blob I get using your plugin. I tried to convert the blob to base64 and then to Uint8Array but it doesn't seems to work, conversion never ended using the DataConverterHelper you provide.

Please, can you tell me if I'm thinking the right way or if I need to change my approach ?

**Comment by qrclip** [comment]
Hello,

So you have a big blob of 500 MB? Or is it already a blob that is part of a large file?

Thanks for trying to use the plugin!

**Comment by qrclip** [comment]
The problem you face is that iOS crashes when you download the file in a big blob... at about 120 MB. The only way is to download the file in chunks and append each chunk to a file. To save, you can use the plugin or the capacitor "Append to File" file system, but that will be slower.
But again, the hardest part is finding a way to download the file in chunks. I've here an example of using fetch and range (it works with S3).
https://github.com/qrclip/capacitor-file-chunk/blob/main/demo/src/app/download01/download-fetch-range.ts

With iOS you can download the 500 MB into memory?

**Comment by berenger-germain-sams** [comment]
Hi,

sorry for the delay, I'm in France ^^.

I already have a big blob of 500MB that I downloaded at once (on Android and iOS). My problem comes when I try to write it in the storage of the device. With capacitor-blob-writer plugin, it works like a charm on Android but failed with a Webkit error on iOS (Error #2 here https://github.com/diachedelic/capacitor-blob-writer#ios-iphone-6)

I understand that I need to have a Uint8Array in order to use your plugin (according to the demo and the method signature) but after the download, I have a Blob object (... of 500MB) so I try to convert it into a Uint8Array and it does not work on Android or iOS. When calling  DataConverterHelper.ConvertBlobToBase64(blob) nothing happens.

Thanks a lot for your help, I really appreciate ;)

**Comment by qrclip** [comment]
Happy to help. So if you already have a blob, you just need to read it in chunks, for example with this function:

`async function readBlobSlice(tBlob: Blob, tOffset: number, tSize: number): Promise<Uint8Array> {
    const tSlice = blob.slice(tOffset, tOffset + tSize);
    const tArrayBuffer = await tSlice.arrayBuffer();
    return new Uint8Array(tArrayBuffer);
}`

Then write(append) each chunk to the file. Each chunk should be small, if you use the capacitor file system, make the chunks about 10 MB, convert them to base64 (now it works) and attach them to a file. If you use the plugin, it'll be much faster. You can try both, because once you have the uint8array, you can also use the plugin. Maybe you try the two approaches.

If you need further help, feel free to ask me. Thanks!

**Comment by berenger-germain-sams** [comment]
I will try both propositions and I will be back to you :)

**Comment by berenger-germain-sams** [comment]
OK, so I implemented both solutions you described and choosed to keep the one with the plugin as it is much faster :) Next week, I will implements the reading of the file and will give you a feedback about my implementation.

I really want to thank you one more time for your help :)

See you!

**Comment by qrclip** [comment]
On iOS, using the plugin has one drawback: if you lock the phone or switch to another app, it stops working. On Android, it takes longer. With the capacitor file system, it just stops and restarts as soon as the app is in the foreground. Reading and writing with the Capacitor File Bridge (converting to base64) isn't extremely slow on Android, but it's with iOS is workable. For appending you can use the Capacitor file system, but you can't read in chunks. For that you can use readFileChunk from the plugin that uses the Capacitor bridge (without localhost) so it doesn't stop. It behaves exactly like the FileSystem. You can use one method for Android and the other for iOS. Just be careful how it works in the background.
Have a great weekend.

**Comment by berenger-germain-sams** [comment]
Hello there,

back to work with a clear mind, I realized that you were right when you asked me:

> With iOS you can download the 500 MB into memory?

😅 Well, I can download it, yes... But I can't handle it! When I try to chunk it, the app crashes (sometime at the first chunk, some other times after 4 or 5 chunks...) I didn't see it earlier because I was developing on smaller files to not wait too long 😖

So I decided to review my whole approach and I implemented it the way you recommend: I download the file chunk by chunk (my Apache server was already able to handle that) and append them as I go along. And tada! It works like a charm. On iOS and Android.

Now I got some problems reading the video. On Android, a simple `Filesystem.getUri()` with `Capacitor.convertFileSrc(uri)` on the HTML `<video>` element works correctly. On iOS, it seems that the video never starts playing. I'm going to explore that soon and will be back to you again to give some news.

**Comment by qrclip** [comment]
Good to know. Now you can easily download large files without running out of memory. As for reading the video, it becomes problematic when that method (Filesystem.getUri()) doesn't work on iOS. I've never had to deal with videos, so I can't help much in this area. Maybe you can try playing the video chunk by chunk? I don't know how hard that will be. First the video format has to allow it, then the transition from chunk to chunk will be the hardest part I think. Let me know how it went. Thanks for your feedback!

**Comment by berenger-germain-sams** [comment]
Hi,

I finally succeeded to achieve my whole feature! So these are some feedbacks :)

The `Filesystem.getUri()` seems to works well on iOS but the system webview had some problems to read big local files. According to this thread (https://github.com/ionic-team/capacitor/issues/5420#issuecomment-1161761109) since capacitor 4.0, it seems that the WKWebview is able to read a local file by range. I can see it working well in my app.

The reading problem I encountered seems to be a well known Iphone problem and when I googled "video can't play on IPhone", I find a lot of issues even on recent devices. So I realized a little test: I captured a video from the camera and tried to play it from the system gallery: it could not play! By just restarting the phone, I was able to play the video I captured and the video downloaded in the app! 🤯

So my decision was to control the reading of the video and advise the user when the video seems to not playing correctly. I published a beta version of the app on both Android and iOS and ask my customer to test the new video feature with multiple devices. For now, all tests were successful even on older devices.

Thanks a lot (again) for your support :)

Have a nice day!
